unit_tests/UnitTestSuppAlgDataSharing.C

```diff
@@ -169,46 +171,45 @@ public:
 
   std::vector<SuppAlg*> suppAlgs_;
   sierra::nalu::ElemDataRequests dataNeededByKernels_;
+  sierra::nalu::FieldManager fieldManager;
 
 private:
   stk::mesh::BulkData& bulkData_;
 };
 
 TEST_F(Hex8Mesh, supp_alg_data_sharing)
 {
-  ScalarFieldType& nodalScalarField = meta->declare_field<ScalarFieldType>(
-    stk::topology::NODE_RANK, "nodalScalarField");
-  VectorFieldType& nodalVectorField = meta->declare_field<VectorFieldType>(
-    stk::topology::NODE_RANK, "nodalVectorField");
-  TensorFieldType& nodalTensorField = meta->declare_field<TensorFieldType>(
-    stk::topology::NODE_RANK, "nodalTensorField");
-  ScalarFieldType& elemScalarField = meta->declare_field<ScalarFieldType>(
-    stk::topology::ELEM_RANK, "elemScalarField");
-  VectorFieldType& elemVectorField = meta->declare_field<VectorFieldType>(
-    stk::topology::ELEM_RANK, "elemVectorField");
-  TensorFieldType& elemTensorField = meta->declare_field<TensorFieldType>(
-    stk::topology::ELEM_RANK, "elemTensorField");
-
   const stk::mesh::Part& wholemesh = meta->universal_part();
+  const stk::mesh::PartVector parts(1, &meta->universal_part());
+  TestAlgorithm testAlgorithm(*bulk);
 
-  stk::mesh::put_field_on_mesh(nodalScalarField, wholemesh, nullptr);
-  stk::mesh::put_field_on_mesh(nodalVectorField, wholemesh, 4, nullptr);
-  stk::mesh::put_field_on_mesh(nodalTensorField, wholemesh, 3, 3, nullptr);
-
-  stk::mesh::put_field_on_mesh(elemScalarField, wholemesh, nullptr);
-  stk::mesh::put_field_on_mesh(elemVectorField, wholemesh, 8, nullptr);
-  stk::mesh::put_field_on_mesh(elemTensorField, wholemesh, 2, 2, nullptr);
+  ScalarFieldType* nodalScalarField =
+    testAlgorithm.fieldManager.register_field<ScalarFieldType>(
+      "nodalScalarField", parts);
+  GenericFieldType* nodalVectorField =
+    testAlgorithm.fieldManager.register_generic_field(
+      "nodalGenericField", parts, 0, 4);
+  TensorFieldType* nodalTensorField =
+    testAlgorithm.fieldManager.register_field<TensorFieldType>(
+      "nodalTensorField", parts);
+  ScalarFieldType* elemScalarField =
+    testAlgorithm.fieldManager.register_field<ScalarFieldType>(
+      "elemScalarField", parts);
+  GenericFieldType* elemVectorField =
+    testAlgorithm.fieldManager.register_generic_field(
+      "elemVectorField", parts, 0, 8);
+  GenericFieldType* elemTensorField =
+    testAlgorithm.fieldManager.register_generic_field(
+      "elemTensorField", parts, 0, 4);
 
   fill_mesh("generated:10x10x10");
 
-  TestAlgorithm testAlgorithm(*bulk);
-
   // TestSuppAlg constructor says which data it needs, by inserting
   // things into the 'dataNeededByKernels_' container.
 
   SuppAlg* suppAlg = new TestSuppAlg(
```

Comment:
This is fine for an intermediate step, but ideally I'm imagining we would just pass a `FieldManager` into `TestAlgorithm`'s ctor and do all the initialization there.

Reply:
If TestAlgorithm should be able to work with only a FieldManager and not a whole stk::mesh::BulkData then I'll try to pass just a FieldManager.  The BulkData is used quite a bit and it is not obvious that FieldManager is all that is needed.  Maybe after looking at the functions that use the BulkData, I will see how FieldManager could be used instead.

Reply:
I was thinking of just adding an argument to the constructor of the TestAlgorithm so it takes the BulkData and a FieldManager. My main thought is that all the field registration that a given algorithm needs should be encapsulated in the Algorithm itself.

FWIW FieldManager has a [reference](https://github.com/Exawind/nalu-wind/blob/58ca0ce7a3deee179ec25b999ac4b60a2b8a945d/include/FieldManager.h#L44) to stk::mesh::MetaData and BulkData can be populated from there. Right now it is private, but we could expose it if that makes sense. If we find that we are always passing FieldManager and Meta/BulkData together then I would say we should expose it.

Reply:
I changed the little test algorithm, `TestSuppAlg`, to register its own fields on the field manager during construction.  There was a change to the mesh constructor since the mesh constructor finalizes the MetaData to exclude new fields. I just moved the mesh constructor after the algorithm constructor, but not sure how that will change in the general case.

Reply:
Okay that's good to know. Something to be aware of.  I suspect that for the main code it won't really make an impact since we register fields all over the place.  mesh finalize is a specific call in Realm IIRC.